src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java

```diff
@@ -51,59 +54,107 @@
+                    // directly after the first byte is received is due to the initialization process of the stream and
+                    // belongs at position 0 (i.e. default background/foreground colors).
+                    if (inCount == 1) {
+                        inCount = 0;
+                    }
+                    LOGGER.log(Level.FINEST, "emitting {0} @{1}/{2}", new Object[] { html, inCount, text.getText().length() });
+                    text.addMarkup(inCount, html);
+                    if (inCount != lastPoint) {
+                        lastPoint = inCount;
+                        int hide = inCount - outgoing.getCount() - adjustment;
+                        // If openTags is not empty, but there are no escape sequences directly on this line, or if we
+                        // are emitting closing tags when closing the stream, there is nothing to hide.
+                        if (hide != 0) {
+                            LOGGER.log(Level.FINEST, "hiding {0} @{1}", new Object[] { hide, outgoing.getCount() + adjustment });
+                            text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<span style=\"display: none\">", "</span>");
+                            adjustment += hide;
+                        }
                     }
                 }
+                public void emitHtmlDirect(String html) {
+                    text.addMarkup(incoming.getCount(), html);
+                }
             }
             EmitterImpl emitter = new EmitterImpl();
-            CountingOutputStream incoming = new CountingOutputStream(new AnsiHtmlOutputStream(outgoing, colorMap, emitter));
-            emitter.incoming = incoming;
-            try {
-                byte[] data = s.getBytes(charset);
-                for (int i = 0; i < data.length; i++) {
-                    // Do not use write(byte[]) as offsets in incoming would not be accurate.
-                    incoming.write(data[i]);
+            // We need to reopen tags that were still open at the end of the previous line so the stream's state is
+            // correct in case those tags are closed in the middle of this line.
+            try (AnsiHtmlOutputStream ansiOs = new AnsiHtmlOutputStream(outgoing, colorMap, emitter, openTags);
+                    CountingOutputStream incoming = new CountingOutputStream(ansiOs)) {
+                emitter.incoming = incoming;
+                // We need to write one UTF-16 code unit at a time so that byte offsets match Java character offsets when inserting HTML.
+                for (int i = 0; i < s.length(); i++) {
+                    // TODO: Do we need to switch to using code points to support Unicode characters outside of the BMP?
+                    char c = s.charAt(i);
+                    if ((c & 0xFF00) != 0) {
+                        incoming.write(new byte[] { (byte) ((c & 0xFF00) >> 1), (byte) (c & 0x00FF) });
```

Comment:
`(byte) (c & 0x00FF)` is equivalent to `(byte) c`, maybe better to switch, although the explicitness might help future readers.

Reply:
Also I guess this is totally wrong for some charsets, working on fixing it.

Reply:
Fixed by c1d7aab.